### PR TITLE
Introducing static type checking

### PR DESCRIPTION
**Pull request by hummusonrails**
This is a first take at introducing static type checking to the SDK. This pull request does the following:

* Adds the two Sorbet gems to the Gemfile
* Initializes Sorbet by introducing the Sorbet folder structure `./sorbet/...`
* Adds `# typed: ...` sigils to each file
* Turns on `typed` to `strict` for `.lib/nexmo/sms.rb`
* Adds signature methods to the SMS class
* Comments out the private class definitions for `Nexmo::Logger` because the return for `@logger.warn` could not be accessed in the private interface
* Moved the `#request` method in `Nexmo::Namespace` to the public interface because it was unreachable in the type check.

### CODE REVIEW

**Comment by hummusonrails** [comment]
@timcraft, Hope you had a good weekend. This is a first take at introducing static type checking into the gem, first in the SMS class. It necessitated some design choices and would love your thoughts on this. 

I'm also going to tag @fabianrbz in this draft PR as another resident Rubyist.

**Comment by timcraft** [comment]
> It necessitated some design choices and would love your thoughts on this.

Private interfaces are private for a reason. If a development tool can't handle private methods or classes then I'd say it's the tool that needs fixing, not the interface that needs making public.

What benefit does this actually add? Did you find any bugs from adding those type definitions? Is there something significant you can do with this that you can't before?

Is it possible to configure sorbet to ignore tests and dev dependencies?

**Comment by hummusonrails** [comment]
> > It necessitated some design choices and would love your thoughts on this.
> 
> Private interfaces are private for a reason. If a development tool can't handle private methods or classes then I'd say it's the tool that needs fixing, not the interface that needs making public.

I agree. It is very possible that there is something I missed in addressing the issue, rather than making the `#request` method public. @fabianrbz, might you have any thoughts on that? 
> 
> What benefit does this actually add? Did you find any bugs from adding those type definitions? Is there something significant you can do with this that you can't before?

That's a good question. I believe the benefit is similar to adding more thorough testing against edge cases. It ensures that the data types we expect is what is being used. Besides the issue mentioned above, which is likely there is a solution that I have not realized yet, its implementation is relatively painless, and the benefit of more of a 'curated' development experience is worth it, in my opinion.

> Is it possible to configure sorbet to ignore tests and dev dependencies?

I believe, but I could be wrong, that all files must have a `# typed: ...` sigil at the top. A designation of `false` means that Sorbet completely ignores that file; `true` means that it checks for code quality, and `strict` means every method must have a method signature. It is up to you to specify what level you want for each file in your code.

**Comment by hummusonrails** [comment]
Moved the `#request` method back under `private`, but redefined `private` to `protected` as the `SMS` class is inheriting from the `Namespace` class, so I think `protected` matches for our purposes more than `private`. Doing so as well resolved the earlier referenced error immediately as well. Thanks to @slorello89 for the suggestion.

File: sorbet/rbi/gems/addressable.rbi [hunk 1]
```diff
@@ -0,0 +1,198 @@
+# This file is autogenerated. Do not edit it by hand. Regenerate it with:
```
Comment (mheap) [inline sorbet/rbi/gems/addressable.rbi:1]: Should the `sorbet` directory be committed?

Comment (hummusonrails) [reply sorbet/rbi/gems/addressable.rbi:1]: Yes, according to the docs, all files generated in the `sorbet` directory should be committed, see Step 6 in https://sorbet.org/docs/adopting.

**Comment by timcraft** [comment]
> I believe, but I could be wrong, that all files must have a `# typed: ...` sigil at the top. A designation of `false` means that Sorbet completely ignores that file; `true` means that it checks for code quality, and `strict` means every method must have a method signature. It is up to you to specify what level you want for each file in your code.

Yes, but that's not what I meant. That sorbet directory includes a lot of code which isn't related to this repo, the addressable.rbi file for example. Presumably the goal of this PR is to type check the code in lib, not to type check the tests, or type check third party dependencies?

Adding 37k loc to type check 1k loc seems a bit ridiculous. Is there a config option in sorbet that can be used to ignore the test files (and any gems used by the tests)?

**Comment by dragonmantank** [comment]
To tack on from our 1:1 with @timcraft , we can ignore some of the folders through sorbet to help cut down on the amount of cruft in the `sorbet/` directory, so we can probably cut down on sorbet looking at tests and any vendors: https://sorbet.org/docs/cli#including-and-excluding-files

**Comment by hummusonrails** [comment]
@timcraft, after further exploration the auto-generated initialization files are non-negotiable. The link @dragonmantank referenced is for specifying which files to check during a runtime check, not during initialization. Having spent a couple of days further looking into this, we're ready to move forward and include this in the 6.3.0 release. The auto-generated files are maintained by the gem, no person touches them or is expected to, so let's go ahead and add this to the upcoming release. Thank you, hope you have a great weekend! 🙏

**Comment by timcraft** [comment]
> The auto-generated files are maintained by the gem, no person touches them

If I understand how it works correctly they need to be updated whenever static type checking is added to a file? So they need to be maintained indirectly, albeit not editing the files by hand.

Just looking at this in more detail:

* The `sorbet/rbi/hidden-definitions/errors.txt` file can be ignored. If you add that path to `.gitignore` that'll cut out 9,517 lines from the diff.

* Can you move the sorbet-runtime require to `lib/nexmo.rb` after the zeitwerk require? That's where runtime dependencies should be loaded (avoids needing to add it to every file).

* The sig for Nexmo::SMS#unicode? looks incorrect, the argument should be a string not nil. I'm pretty sure you can remove the `T.nilable` from there. Also it helps to format whitespace consistently, prefer `sig {` to `sig{` for example in that case.

* Instead of using `T.let` for every reference to `@logger` I think it could be simpler to eliminate the instance variable, reference `config.logger` directly, and add a protected accessor for the `config` attribute. It doesn't need to be an instance variable, that's just how it was written.

* The `private_constant` calls exist in order to clearly mark "internal" code, and those constants don't show up in the generated documentation. If sorbet doesn't support that usage then the calls to `private_constant` need to be removed (not simply commented out), and they need to be marked as private so that YARD doesn't generate documentation for them.

Let me know if any of those don't make sense and we can discuss.

**Comment by hummusonrails** [comment]
Hi @timcraft - I see a bunch of work has been done on this just now. Is this in good shape for next week? If not, what are we missing?

**Comment by timcraft** [comment]
@benhayehudi I've rebased on master and fixed a number of outstanding issues.

I'll incorporate into v6.3.0 as discussed.
